### ISSUE TITLE
Introspection information

### ISSUE THREAD

**Issue by spk121** [spk121/guile-gi#24]
To the extent possible for generated functions, methods, classes, and constants, make the docstrings that appear in the typelib file available as introspection information and at the REPL.

This will likely require modification of upstream GOOPS and the Guile REPL, since GOOPS methods don't appear to support docstrings, and the best way to provide docstrings to overloaded procedures is not decided.

**Comment by LordYuuma** [comment]
Actually, we should be able to set a procedure property to the documentation string and define a GOOPS method that fetches this property.

**Comment by spk121** [comment]
That would be useful.

But if I could have exactly what I want....
Imagine a pair of overloaded functions
```
(define-method (blammo (x <string>)) "docstring" ...)
(define-method (blammo (x <integer>)) "docstring2" ...)
```
From the REPL, I hope to get to something like
```
scheme@(guile-user)> ,describe blammo
  1. (blammo (x <string>))
   "docstring"
  2. (blammo (x <integer>))
   "docstring2"
scheme@(guile-user)> ,apropos blammo
  (guile-user): blammo  #<<generic> (x <string>)>
  (guile-user): blammo  #<<generic> (x <integer>)>
```

**Comment by LordYuuma** [comment]
Well, I'm not sure about ,describe, but there is `(oop goops describe)`.

    scheme@(guile-user)> (use-modules (gi) (oop goops describe))
    scheme@(guile-user)> (use-typelibs ("GObject" "2.0"))
    scheme@(guile-user)> (describe <GObject>)
    <GObject> is a class. It's an instance of <class>
    Superclasses are:
        <GFundamental>
    Directs slots are:
        ref
        unref
    Directs subclasses are:
        <GTypeModule>
        <GInitiallyUnowned>
        <GBinding>
    Class Precedence List is:
        <GObject>
        <GFundamental>
        <object>
        <top>
    Class direct methods are:
        Method #<<method> (<GObject> <top>) 55f10d214240>
    	     Generic: signal-stop-emission-by-name
    	Specializers: <GObject> <top>
        Method #<<method> (<GObject> <top> <top>) 55f10d214300>
    	     Generic: signal-stop-emission
    	Specializers: <GObject> <top> <top>
        ...

**Comment by LordYuuma** [comment]
After further inspecting Guile internals, it seems as if `,describe` reads procedure and object properties. Hence, if you do

    (define-method (blammo (x <string>)) "docstring" ...)
    (define-method (blammo (x <integer>)) "docstring2" ...)
    (set-object-property blammo 'docstring "1. (blammo (x <string>))\n \"docstring\"\n...")

You should get what you want. You can also call this from C using `scm_set_object_property_x`. Perhaps we should write a wrapper `scm_append_object_documentation_x` in `gig_util`.

**Comment by spk121** [comment]
I did verify that you can generate docstrings for generics by entering the information in a `guile-procedures.txt`.   There isn't any facility for making a docstring for the specific method (that'll require upstream work) but one could document all of the same-named generics in one monolithic docstring.

Making this would would be a bit of a hack, since it would require parsing a `.gir` and a typelib to generate the docstrings, and then save the `guile-procedures.txt` in a location that `(@@ (ice-9 documentation) documentation-files)` will search for them.

**Comment by LordYuuma** [comment]
What a coincidence. We already need the same for the offline documentation. I'll figure out some XML merging for #23 to pull in all information, and once that is done, we can write some "backends" such as `->guile-procedures.txt`, `->docbook` and `->brainfuck-executable`.

It seems `documentation-files` is a variable, so perhaps we can store documentation in non-standard locations by modifying it. Even hackier than what you suggested, I know.

**Comment by LordYuuma** [comment]
`->guile-procedures.txt` can now handle doc strings. Twas a weird experience recoding, but it's worth it.

**Comment by LordYuuma** [comment]
Given the current state of `->guile-procedures.txt`, is this issue resolved, or do we still lack something essential?

**Comment by spk121** [comment]
Looks good to me.

**Comment by LordYuuma** [comment]
Resolved after #48.
